crates/rethnet_eth/src/remote/client.rs

```diff
@@ -226,67 +300,148 @@ impl RpcClient {
+        let request = self.serialize_request(method_invocation)?;
 
-        self.send_request_body_with_cache(&request.request, cache_key)
-            .await
-            .and_then(|response| Self::extract_response(request, response))
+        let result = if let Some(cached_response) = self.try_from_cache(cache_key.as_ref()).await? {
+            serde_json::from_value(cached_response).expect("cache item matches return type")
+        } else {
+            let result: T = self
+                .send_request_body(&request)
+                .await
+                .and_then(|response| Self::extract_result(request, response))?;
+
+            if let Some(cache_key) = cache_key {
+                self.write_response_to_cache(&cache_key, &result).await?;
+            }
+
+            result
+        };
+        Ok(result)
     }
 
     // We have two different `call` methods to avoid creating recursive async functions as the
     // cached path calls `eth_chainId` without caching.
-    async fn call_without_cache<T>(&self, input: &MethodInvocation) -> Result<T, RpcClientError>
-    where
-        T: for<'a> serde::Deserialize<'a>,
-    {
-        let request = self.serialize_request(input);
+    async fn call_without_cache<T: DeserializeOwned>(
+        &self,
+        method_invocation: MethodInvocation,
+    ) -> Result<T, RpcClientError> {
+        let request = self.serialize_request(method_invocation)?;
 
-        self.send_request_body(&request.request)
+        self.send_request_body(&request)
             .await
-            .and_then(|response| Self::extract_response(request, response))
+            .and_then(|response| Self::extract_result(request, response))
     }
 
+    /// Returns the results of the given method invocations.
     async fn batch_call(
         &self,
-        inputs: &[MethodInvocation],
-    ) -> Result<BatchResponse, RpcClientError> {
-        let cache_key = CacheableMethodInvocations::try_from(inputs)
-            .ok()
-            .and_then(|v| v.cache_key());
+        method_invocations: &[MethodInvocation],
+    ) -> Result<VecDeque<serde_json::Value>, RpcClientError> {
+        let ids = self.get_ids(method_invocations.len() as u64);
 
-        let request_strings: Vec<String> = inputs
+        let cache_keys = method_invocations
             .iter()
-            .map(|i| self.serialize_request(i).request)
-            .collect();
+            .map(|m| {
+                CacheableMethodInvocation::try_from(m)
+                    .ok()
+                    .and_then(|v| v.cache_key())
+            })
+            .collect::<Vec<_>>();
 
-        let request_body = format!("[{}]", request_strings.join(","));
+        let mut results: Vec<Option<serde_json::Value>> =
+            Vec::with_capacity(method_invocations.len());
+        for cache_key in cache_keys.iter() {
+            let result = self.try_from_cache(cache_key.as_ref()).await?;
+            results.push(result)
+        }
```

Comment:
I don't know of any, but I figured I'd ask, in case you know: is there a way to do async code in functional code, such that you can do: `.iter().map(..).collect::<Future<Result<Vec<Option<serde_json::Value>>>>`?

Reply:
It'd be possible with [`futures::stream`](https://docs.rs/futures/latest/futures/stream/index.html), but imo it's more difficult to read than the imperative version. Would make sense though if we wanted to parallelize the disk reads.

Reply:
I thought it would make some of this code cleaner with regard to the pre-allocations as those can be `filter_map`ped and `collect`ed respectively

Reply:
Ok, want me to give it a try? (We need the `results` to be be as long as as the input `method_invocations` as we'll replace the `None` values with the results from the remote node based on the indices.

Reply:
If it's trivial to do, give it a try :)